lib/std/time.zig

```diff
@@ -127,85 +129,31 @@ pub const s_per_hour = s_per_min * 60;
 pub const s_per_day = s_per_hour * 24;
 pub const s_per_week = s_per_day * 7;
 
-/// A monotonic high-performance timer.
-/// Timer.start() must be called to initialize the struct, which captures
-/// the counter frequency on windows and darwin, records the resolution,
-/// and gives the user an opportunity to check for the existnece of
-/// monotonic clocks without forcing them to check for error on each read.
-/// .resolution is in nanoseconds on all platforms but .start_time's meaning
-/// depends on the OS. On Windows and Darwin it is a hardware counter
-/// value that requires calculation to convert to a meaninful unit.
+/// A high-performance timer that tries to be monotonically increasing.
```

Comment:
Would require a bit of synchronization overhead for single-threaded use cases. The current alternative is to [enforce it locally](https://github.com/coilhq/tigerbeetle/blob/main/src/time.zig) instead of globally. The overhead might not make much of a difference given we're already converting timestamps to nanoseconds (mul + div, possibly "expensive") for each read()?

Reply:
Ah I see your point. Well we already have the concept of creating an instance with the start() function. That instance could store the value to enforce it locally. I suggest we do that, and keep the API simpler. If `nanoTimeStamp` has to be deleted, that's fine with me.

Reply:
@kprotty is correct, `nanoTimeStamp` can't be monotonic, since the system time can change forward and backward for a number of reasons.

With all the issues in `nanoTimeStamp`, I'd prefer it go away entirely, so that it doesn't lead people down a path of bugs.

Reply:
> For Timer to enforce monotonic values, read() would have to take in `*Timer` instead of `Timer` (to update the "last timestamp"). Is this fines?

Yes


> Also, to be clear, `nanoTimeStamp` doesn't have to be monotonic. It returns the system time so it has an unfortunate name which makes people use it for measuring elapsed time. Rename it to `systemTimestamp` or nah?

Oops, my mistake. I have another name counter-proposal: `calendarTimestamp`

Reply:
For `calendarTimestamp`, we would lose the other millis/second resolutions unless we had like `calendarMillisTimestamp`. Also, "calendar" feels off given my recent understanding about leap seconds, unless the time returned is time-zone specific. I've been talking with others on the discord and came up with counter-counter-proposal (also API wise):

```zig
/// Get measurements of time relative to the unix epoch (jan 1 1970)
/// Precision depends on OS.
/// For capturing date time. Use Instant or Timer for capturing elapsed time.
pub const UnixTime = struct {
    pub fn nanoTimestamp() i128
        // windows: GetSystemTimePreciseAsFileTime()
        // darwin: gettimeofday (clock_gettime calls this internally)
        // posix: clock_gettime(CLOCK_REALTIME)
    pub fn timestamp() i64 // seconds version
    pub fn milliTimestamp() i64 // millis version
};

/// Measures amount of time the callers thread or process spent running on cpu
pub const CpuTime = enum {
    thread,
    process,
    
    // measure thread/process cpu time in nanoseconds 
    pub fn read(self: CpuTime) error{Unavailable}!u64
        // windows: GetThreadTimes/GetProcessTimes
        // posix: clock_gettime(CLOCK_[THREAD|PROCESS]_CPUTIME_ID)
};

/// A snapshot of the current high-precision clock value.
/// Mostly monotonic, but is allowed to jump backwards.
/// Can be used to measure elapsed (or wallclock) time. 
pub const Instant = struct {
    timestamp: if (is_darwin or is_windows) u64 else os.timespec,
    
    // Get the current Instant snapshot.
    // Time conversion to nanoseconds happens at since() to keep now() sampling fast.
    pub fn now() error{Unavailable}!Instant
        // windows: QueryPerformanceCounter
        // darwin: mach_continuous_time
        // linux: clock_gettime(CLOCK_BOOTTIME)
        // posix: clock_gettime(CLOCK_MONOTONIC)

    // For fast comparisons without computing time since() below
    pub fn order(self: Instant, other: Instant) std.math.Order 
    
    // Returns the amount time in nanoseconds since the earlier Instant to this Instant.
    // If this Instant snapshot represents a time before earlier Instant, a negative value is returned.
    // If too much time occurred between this Instant and earlier Instant, `error.Overflow` is returned.
    pub fn since(self: Instant, earlier: Instant) error{Overflow}!i64
        // windows: QueryPerformanceFrequency / specialize on 10mhz / safeMulDiv
        // darwin: mach_timebase_info / specialize on numer==denom / safeMulDiv
        // posix: timespecsub()
};

// Same thing as before, but backed by Instant and is ensured to be monotonic
pub const Timer = struct {
    start_time: Instant,
    last_time: Instant,

    pub fn start() error{TimerUnsupported}!Timer:
        const s = Instant.now() catch return error.TimerUnsupported;
        return Timer{ .start_time = s, .last_time = s };
        
    pub fn reset(self: *Timer) void:
        self.start_time = self.getTimestamp()
        
    pub fn read(self: *Timer) u64:
        return self.toDuration(self.getTimestamp())
    
    pub fn lap(self: *Timer) u64:
        const current_time = self.getTimestamp();
        defer self.start_time = current_time;
        returns self.toDuration(current_time);
    
    // reads timestamp assuming that it cannot fail
    // makes sure the timestamp is monotonic
    fn getTimestamp() Instant {
        const current = Instant.now() catch unreachable;
        if (self.last_time.order(current) == .lt) self.last_time = current;
        return self.last_time 

    fn toDuration(current: Instant) u64 {
        assert(self.start_time.order(current) != .lt);
        return @intCast(u64, current.since(self.start_time) catch unreachable)
};
```